258 create assignment

rylan-geykhman opened the pull request:
<!--------------------------------
Before you submit your PR:
1. Do you have any console.logs for testing purposes? Remove them please!
2. Does your code actually work? Test all of your changes one more time!
3. Fill out the section below. It helps us review your PR because I ain't reading allat lil bro
4. Delete these instruction comments (optional)
--------------------------------->

<!--------------------------------------------------------------------------------------------------------------
Summarize any changes your PR includes. If they are related to an issue, tag it at the end.

Examples:
- Added yoda to home page
- Created yoda clicker component (#69420)
- Fixed bug where yoda does not spawn after 5 clicks (#80085)
-------------------------------------------------------------------------------------------------------------->

### Changes

- Create Assignment page now exists
- Form to create said assignment to whatever courseId
- Assignment will show up on the course page.

rylan-geykhman committed:
function change

rylan-geykhman committed:
commit

rylan-geykhman committed:
Merge branch 'dev' into 258-create-assignment

rylan-geykhman committed:
commit

rylan-geykhman committed:
commit

rylan-geykhman committed:
Merge branch 'dev' into 258-create-assignment

rylan-geykhman committed:
cleaned up code, matched api inputs

- make sure inputs work
- test to make sure it works
- comments to look into certain things
- eta friday

rylan-geykhman committed:
timerStyle quickly removed

rylan-geykhman committed:
question select

- maybe a working model?
- try it out

rylan-geykhman committed:
dropdown for questions

- still need to make an alternative for datetime-local so it supports all browsers
- fetching courseID and not even asking for that as an input
- 0 should mean infinite for both time allotted and attempts allowed? confirm needed

rylan-geykhman committed:
dropdown fixes, other fixes

- dropdowns now smoother via batching
- date and time split into two inputs, put together as one and converted to iso string
- 401 Unauthorized error when trying to create, need to look into that

rylan-geykhman committed:
commit

rylan-geykhman committed:
the final commit

rylan-geykhman committed:
pr stuff

kennething committed:
fix eslint

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+      <div class="relative">
+        <label for="guaranteedQuestions">Guaranteed Questions</label>
+        <div class="mt-1 block w-full cursor-pointer rounded-md border-gray-300 px-3 py-2 shadow-sm focus:ring-blue-500 sm:text-sm" @click="toggleGuaranteedDropdown">
+          <span v-if="guaranteedQuestions.length > 0">{{ guaranteedQuestions.length }} Selected</span>
```
should be a ternary instead of v-if/v-else

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+          <span v-else>Select Questions</span>
+        </div>
+        <div v-if="guaranteedDropdownOpen" class="absolute z-10 mt-1 max-h-60 w-full overflow-y-auto rounded-md border border-gray-300 bg-white shadow-lg">
+          <div v-for="n in getGuaranteedCurrentBatch()" :key="'guaranteed-' + n" class="flex items-center space-x-2 px-3 py-2 hover:bg-gray-100">
```
1. most descriptive variable name in the west
2. key is never seen anywhere by anyone except for vue, it doesn't need to make sense to humans, it just needs to be a unique value (just `:key="n"` will suffice)

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+    <form class="space-y-4" @submit.prevent="createAssignment">
+      <input v-model="name" type="text" placeholder="Assignment Name" class="input-field" required />
+      <!--courseID probably auto filled in?-->
+      <input v-model.number="courseID" type="number" placeholder="Course ID" class="input-field" required />
```
1. why would they need to input the course id themselves? we already have it :skull:
2. what is this `input-field` class for

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+
+<script setup lang="ts">
+const name = ref("");
+const courseID = ref();
```
needs an explicit type or a default value for implicit type

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+const randomQuestions = ref<number[]>([]);
+const dueDateDate = ref("");
+const dueDateTime = ref("");
+const numOfQuestions = ref();
```
this should always be a number?

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+const dueDateTime = ref("");
+const numOfQuestions = ref();
+const lateSubmissions = ref(false);
+const timeAllotted = ref();
```
this should always be a number?

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+const numOfQuestions = ref();
+const lateSubmissions = ref(false);
+const timeAllotted = ref();
+const attemptsAllowed = ref();
```
this should always be a number?

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+
+function getISODateTime() {
+  if (!dueDateDate.value || !dueDateTime.value) return null;
+  const dateTimeString = `${dueDateDate.value}T${dueDateTime.value}:00`;
```
the `Date()` constructor can take a string date just fine, you don't need to
1. manually format the string as an ISO string
2. create a new date from the manually formatted ISO string
3. turn the new date created from the manually formatted ISO string, into an ISO string

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  return isoDate;
+}
+
+function getGuaranteedCurrentBatch() {
```
this function and the function below can and should definitely be turned into 1 function but with a parameter

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  return allQuestions.slice(start, end);
+}
+
+function getRandomCurrentBatch() {
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  return allQuestions.slice(start, end);
+}
+
+function toggleGuaranteedDropdown() {
```
this is a 1 line function and only used in the template, this is unnecessary

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  guaranteedDropdownOpen.value = !guaranteedDropdownOpen.value;
+}
+
+function toggleRandomDropdown() {
```
this is a 1 line function and only used in the template, this is unnecessary

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  randomDropdownOpen.value = !randomDropdownOpen.value;
+}
+
+function goToNextGuaranteedPage() {
```
this function and the function below can and should definitely be turned into 1 function but with a parameter

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  }
+}
+
+function goToPreviousGuaranteedPage() {
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  }
+}
+
+function goToNextRandomPage() {
```
this function and the function below can and should definitely be turned into 1 function but with a parameter

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+  }
+}
+
+function goToPreviousRandomPage() {
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+const randomCurrentPage = ref(1);
+
+function getISODateTime() {
+  if (!dueDateDate.value || !dueDateTime.value) return null;
```
1. ~~most efficient guard clause in the east (checking the same value twice?)~~
2.  dont return null in a guard clause, return nothing (undefined)
`return;`

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+    if (!dueDate) throw new Error("Invalid date or time");
+    await submitCreateAssignment(
+      name.value,
+      Number(courseID.value),
```
why are we using a number constructor on something that is always a number

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+      guaranteedQuestions.value,
+      randomQuestions.value,
+      dueDate,
+      Number(numOfQuestions.value),
```
why are we using a number constructor on something that is always a number

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+      dueDate,
+      Number(numOfQuestions.value),
+      lateSubmissions.value,
+      Number(timeAllotted.value),
```
why are we using a number constructor on something that is always a number

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,168 @@
+      Number(numOfQuestions.value),
+      lateSubmissions.value,
+      Number(timeAllotted.value),
+      Number(attemptsAllowed.value)
```
why are we using a number constructor on something that is always a number

kennething reviewed (changes requested):
do NOT let chatgpt cook lil bro :skull: :skull: :skull:

kennething committed:
Revert "fix eslint"

This reverts commit e5952808fd01cc5bb5291b12b7c0792cc7ad1d63.

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+<template>
+  <div class="mx-auto max-w-lg rounded-lg p-6 shadow-md">
```
`mx-auto` is unnecessary, thats default behavior

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+  <div class="mx-auto max-w-lg rounded-lg p-6 shadow-md">
+    <h2 class="mb-4 text-2xl font-semibold">Create Assignment</h2>
+    <form @submit.prevent="createAssignment" class="space-y-4">
+      <input v-model="name" type="text" placeholder="Assignment Name" class="input-field" required />
```
what is this `input-field` class for

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+          <span v-else>Select Questions</span>
+        </div>
+        <div v-if="randomDropdownOpen" class="absolute z-10 mt-1 max-h-60 w-full overflow-y-auto rounded-md border border-gray-300 bg-white shadow-lg">
+          <div v-for="n in getRandomCurrentBatch()" :key="'random-' + n" class="flex items-center space-x-2 px-3 py-2 hover:bg-gray-100">
```
dont need string + n in key

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+      </div>
+      <div class="space-y-2">
+        <label>Due Date: </label>
+        <input v-model="dueDateDate" type="date" class="input-field" required />
```
`input-field` class

kennething replied:
whats the point of the class?

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+      <div class="space-y-2">
+        <label>Due Date: </label>
+        <input v-model="dueDateDate" type="date" class="input-field" required />
+        <input v-model="dueDateTime" type="time" class="input-field" required />
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+        <input v-model="dueDateDate" type="date" class="input-field" required />
+        <input v-model="dueDateTime" type="time" class="input-field" required />
+      </div>
+      <input v-model.number="numOfQuestions" type="number" placeholder="Number of Questions" class="input-field" required />
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+        <span>Allow Late Submissions</span>
+      </label>
+      <!--0 = unlimited time, rest is by minutes? ask about this-->
+      <input v-model.number="timeAllotted" type="number" placeholder="Time Allotted (mins)" class="input-field" required />
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+      </label>
+      <!--0 = unlimited time, rest is by minutes? ask about this-->
+      <input v-model.number="timeAllotted" type="number" placeholder="Time Allotted (mins)" class="input-field" required />
+      <input v-model.number="attemptsAllowed" type="number" placeholder="Attempts Allowed" class="input-field" required />
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+<script setup lang="ts">
+const name = ref("");
+const courseID = ref();
+const guaranteedQuestions = ref<number[]>([]);
```
this isnt an array of questions, its an array of question ids
make variable name more clear

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+const name = ref("");
+const courseID = ref();
+const guaranteedQuestions = ref<number[]>([]);
+const randomQuestions = ref<number[]>([]);
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+
+const allQuestions = Array.from({ length: 5573 }, (_, i) => i + 1);
+const batchSize = 100;
+const guaranteedCurrentPage = ref(1);
```
1-based index? we dont code in baby languages like lua
this is an index, make variable name more clear

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+const allQuestions = Array.from({ length: 5573 }, (_, i) => i + 1);
+const batchSize = 100;
+const guaranteedCurrentPage = ref(1);
+const randomCurrentPage = ref(1);
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,167 @@
+const courseID = ref();
+const guaranteedQuestions = ref<number[]>([]);
+const randomQuestions = ref<number[]>([]);
+const dueDateDate = ref("");
```
make variable name more different from the one below

rylan-geykhman committed:
Merge branch 'dev' into 258-create-assignment

rylan-geykhman committed:
Revert "Merge branch 'dev' into 258-create-assignment"

This reverts commit 72a3f9425d8e41ba2f7799f392fdddda05b79806, reversing
changes made to 947a7902af7e28690b0e9daf7e0149425ae748b7.

rylan-geykhman committed:
Reapply "Merge branch 'dev' into 258-create-assignment"

This reverts commit a557888f9281eedbe856b535887d382ff5e2d20b.

rylan-geykhman committed:
some pr fixes

rylan-geykhman committed:
eslint fixes

rylan-geykhman committed:
lots of changes to make better

- datetime-local (again)
- courseID automatically filled (illusion of choice)
- other fixes

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,153 @@
+        </div>
+        <div class="space-y-2">
+          <label class="flex items-center space-x-2">Due Date: </label>
+          <input v-model="dueDateInput" type="datetime-local" />
```
add required

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,153 @@
+const randomQuestionIDs = ref<number[]>([]);
+const dueDateInput = ref("");
+const dateError = ref("");
+const numOfQuestions = ref<number>(0);
```
dont explicitly type things that are implicitly typed https://kennethng.dev/blog/how-to-write-not-shit-vue-typescript-code#explicit-typing

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,153 @@
+const dateError = ref("");
+const numOfQuestions = ref<number>(0);
+const lateSubmissions = ref(false);
+const timeAllotted = ref<number>(0);
```
^

kennething commented on pages/teacher/course/[courseCode]/create-assignment.vue:
```diff
@@ -0,0 +1,153 @@
+const numOfQuestions = ref<number>(0);
+const lateSubmissions = ref(false);
+const timeAllotted = ref<number>(0);
+const attemptsAllowed = ref<number>(0);
```
^

rylan-geykhman committed:
even more kenneth fixes

kennething reviewed (approved):
LGTM

rylan-geykhman committed:
Merge branch 'dev' into 258-create-assignment
